[WIP] Adding ParametricIntegral class

friyaz opened the pull request:
<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->

#### References to other Issues or PRs
<!-- If this pull request fixes an issue, write "Fixes #NNNN" in that exact
format, e.g. "Fixes #1234" (see
https://tinyurl.com/auto-closing for more information). Also, please
write a comment on that issue linking back to this pull request once it is
open. -->
#19320 

#### Brief description of what is fixed or changed
An object of ParametricIntegral represents integral of a scalar or vector field over a Parametric Region.

#### Other comments
`_bounds_case` function is taken from @prasoon2211  [PR](https://github.com/sympy/sympy/pull/2208/files) with some modification. I have added him as a coauthor for the corresponding commit. 
#### Release Notes

<!-- Write the release notes for this release below. See
https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more information
on how to write release notes. The bot will check your release notes
automatically to see if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->
* vector
    * added class to represent integral of scalar/vector field over a parametric surface.
<!-- END RELEASE NOTES -->

friyaz committed:
Added some tests for ParametricIntegral Class

Upabjojr commented on sympy/vector/integrals.py:
```diff
@@ -0,0 +1,30 @@
+from sympy.vector.coordsysrect import CoordSys3D
+from sympy.integrals import Integral
+
+class ParametricIntegral(Integral):
```
`Integral` inherits associative operator, which itself overloads the `__new__` constructor.

Upabjojr commented on sympy/vector/integrals.py:
```diff
@@ -0,0 +1,30 @@
+    8*pi
+
+    >>> ParametricIntegral(C.j - C.k, ParametricRegion((r, theta), (r*cos(theta), r*sin(theta))))
+    ParametricIntegral(C.j - C.k, ParametricRegion((r, theta), (r*cos(theta), r*sin(theta))))
```
This API looks ok to me.

friyaz committed:
Added suuport for line integrals in ParametricIntegral

friyaz commented on sympy/vector/integrals.py:
```diff
@@ -0,0 +1,79 @@
+    
+    def __new__(cls, field, parametricregion):
+        obj = super().__new__(cls, field, parametricregion)
+        return obj
```
I am not sure how should I make this a subclass of Integral.

Upabjojr replied:
No need to inherit any classes, unless you really want to reuse some code of the upper class.

friyaz commented on sympy/vector/integrals.py:
```diff
@@ -0,0 +1,79 @@
+        if len(coord_sys) > 1:
+            raise ValueError
+        
+        coord_sys = next(iter(coord_sys))
```
Since no coordinate system is associated with the parametric region, we use the base scalars of the field. But if the field does not have any basescalar or vector, we cannot derive coordinate system. This breaks the logic

Example:
```python
>>> p = ParametricRegion(t, (3*t - 2, t + 1), {t: (1, 2)}).
>>> I = ParametricIntegral(C.x + C.y, p)
>>> I.eval()
5*sqrt(10)
>>>
>>> I = ParametricIntegral(1, p) ##We cannot extract voordinate system from 1
StopIteration
```

sylee957 replied:
My opinion is that we should equip base scalars and base vector (fields) to the `ParametricRegion` if we want to use its intrinsic scalars and vector field components.
I don't think that it makes much sense to have vector integral like `ParametricIntegral(C.x*C.i, p)` where C.x is linked to the base scalars of the parametric region and `C.i` is linked to embedding cartesian coordinate system.

friyaz replied:
Agreed. But if we allow the coordsys parameter to be optional and None by default in ParametricRegion class, same situation can arise.

friyaz replied:
@Upabjojr, your thoughts on this?

Upabjojr replied:
Probably it's better to separate vector and parametric integrals. At least for now.

friyaz replied:
I am not sure what you mean here. Do you mean a separate class for handling vector fields? Please explain.
I think it will be better if we associate some default coordinate system with base scalars with a ParamRegion Object?

Upabjojr replied:
What about temporarily ignoring coordinate systems? We can go back to them in the future.

Upabjojr replied:
`ParametricIntegral(C.i, ... )` is a vector, while `ParametricIntegral(C.x, ... )` is a scalar. I would expect different classes for vectors and scalars. For example, vectors should be endowed with methods for vector operations.

friyaz replied:
> I would expect different classes for vectors and scalars. 

I think it will be easier for users if we have a common class for vector and scalar fields. Also, the calculation is very similar in both cases. It just differs whether we take dot product with field or multiply it with a modulus of the normal vector to the surface

If we decide to make separate classes,  we also should have a class that returns an object of ParametricScalarIntegral or ParametricVectorIntegral depending on the nature of the field.

> vectors should be endowed with methods for vector operations

Does it mean applying curl and divergence operators as used in Green's theorem and the Divergence theorem? It would be nice to have them.  but we can easily apply them to the field without any method too.
```python
ParametricIntegral(curl(C.x*C.y*C.i), surface))
```

Upabjojr replied:
What about using class inheritance?

friyaz replied:
Yes, we can make a 'ParametricRegion` class and make two subclasses: `ScalarParametricIntegral` and `VectorParametricIntegral`.
But I think It would be easier if a user can use a single class for both scalar and vector fields.

friyaz replied:
> ParametricIntegral(C.i, ... ) is a vector, while ParametricIntegral(C.x, ... ) is a scalar
How can `ParametricIntegral(C.i, ...)` be a vector? I think it is assumed in most textbooks that the dot product of vector field and position vector of the Parametric region takes place and the result is integrated along the region.

friyaz replied:
Maybe we can have a function `vectorintegrate` which returns the object of ParametricVectorIntegral or ParametricScalarIntegral depending on the nature of the field. 

This can work for other types of integrals too. Suppose we implement a class ImplicitIntegral to calculate the integral over an implicit region. The `vectorintegrate` function can then return an object of ImplicitIntegral object by identifying the region is defined implicitly.

@Upabjojr

Upabjojr replied:
There are various ways to dispatch. Yes, in general the users should not have to choose between the subclass... there should be one function and then the code should decide which class to use.

The point of having subclasses is that the result of the vector integration may be a vector or a scalar, depending on what and how you are integrating.

The vector object should behave like a vector (e.g. have methods of vectors).

friyaz replied:
> The point of having subclasses is that the result of the vector integration may be a vector 

I do not see how this is possible.

friyaz replied:
I think this is possible if the user want to perform cross product of vector field with normal vector instead of the usual dot product.

Upabjojr replied:
Well, integrating velocity over time is a simple example. Though it's probably already handled by the current `Integral` class. But you see, this case could be a case where you may want to subclass `Integral` in order to supply it with the vector methods.

Upabjojr replied:
But indeed, probably we don't need this case, it's pretty rare that some users may wish to use it.

friyaz replied:
Although it is a kind of vector integral, it is not integrating a vector over a region. As you said, the `integrate` function is able to handle such kind of integrals (integral over variables). But if we wish to handle such case, we can allow passing a tuple of variables in place of the parametric region.

friyaz committed:
Added support for surface integrals in ParametricIntegral class

friyaz commented on sympy/vector/integrals.py:
```diff
@@ -0,0 +1,95 @@
+                            (v, lower_v, upper_v))
+            else:
+                result = integrate(parametricfield*normal_vector.magnitude(), (u, lower_u, upper_u), (v, lower_v, upper_v))
+        
```
Changing the order of parameters cuases Integral to fail or get slow.
```python
>>> field = sqrt(C.x**2 + C.y**2)
>>> cone1 = ParametricRegion((u,v), ((2-2*u/3)*cos(v), (2-2*u/3)*sin(v), u), {u: (0, 3), v: (0, 2*pi)})
>>> cone2 = ParametricRegion((u,v), ((2-2*u/3)*cos(v), (2-2*u/3)*sin(v), u), {u: (0, 3), v: (0, 2*pi)})
>>> ParametricIntegral(field, cone1)
# It gets stuck
>>> ParametricIntegral(field, cone2)
8*sqrt(13)*pi/3
```

I think this is because `normal_vector.magnitude` becomes a large expression and Integral is not able to work on it. Using `simplify` on `normal_vector.magnitude` solves the problem.

Upabjojr replied:
`simplify` is very expensive in computational terms. What about some more specific functions (e.g. `expand`)?

friyaz replied:
@Upabjojr It gets slow if I use `expand`.

friyaz replied:
I think `Integrate` function should be able to handle or perform such simplification instead of simplifying them here.

friyaz committed:
Modified ParametricIntegral to handle Surface integrals with interdependent bounds

Co-authored-by: Prasoon Shukla <prasoon92.iitr@gmail.com>

friyaz commented on sympy/vector/integrals.py:
```diff
@@ -0,0 +1,141 @@
+            # Case 1 : Both bounds are constants : can  perform integration in any order
+            # Case 2 : Bounds for u are in terms of v : integrate wrt to u first
+            # Case 3 : Bounds for v are in terms of u : integrate wrt to v first
+            
```
Determining order for integration is necessary when bounds of parameters are interdependent.
Suppose we need to calculate area of a triangle:
```python
>>> integrate(1, (x, 0, 2), (y, 10 - 5*x))
20 - 10*x
>>> integrate(1, (y, 0, 10 - 5*x), (x, 0, 2))
10 # correct answer
```

Upabjojr replied:
Topological sort?

friyaz replied:
Yes, we need to perform a topological sort.
This function generates the graph and performs topological sort. It works on any number of parameters.
```python
    @classmethod             
    def _topological_sort_parameters(cls, parameters, limits):
    
        V = list(parameters)
        E = list()
        
        for p in parameters:
            lower_p = limits[p][0]
            upper_p = limits[p][1]
            
            lower_p = lower_p.atoms()
            upper_p = upper_p.atoms()
            for q in parameters:
                if p == q:
                    continue
                if lower_p.issuperset(set([q])) or upper_p.issuperset(set([q])):
                    E.append((p, q))   
        return topological_sort((V, E), key=default_sort_key)   
```
The only problem with this approach is when bounds are independent of each other, we can perform integration in any order. So if one order fails to calculate the integral, we can reverse the order and perform again. With my previous approach using cases, we can determine whether bounds are independent of each other but it works only for two parameters.

Upabjojr replied:
What about calling the equation solver? The equation solver should be able to find a parametrization somehow.

friyaz committed:
Use topological sort to order parameters

friyaz committed:
Added support for integrating 3 and higher dimension parametric surfaces

Upabjojr commented on sympy/vector/integrals.py:
```diff
@@ -0,0 +1,128 @@
+
+            lower, upper = parametricregion.limits[parameter][0], parametricregion.limits[parameter][1]
+            
+            if isinstance(parametricfield, Vector):
```
Maybe this case should be moved to another class, e.g. `ParametricVectorField`?

friyaz committed:
ParametrcIntegral class can integrate sclar fields free from base-scalars

Upabjojr commented:
Tests are broken.

friyaz committed:
Added vector_integrate function and fixed test cases of ParametricIntegral class

friyaz commented:
@Upabjojr I have fixed them. 
Also, I have added `vector_integrate` function.  I will add the test cases for it in the next PR if it is decided to keep this function.

Upabjojr commented:
It works for me. Thanks.
